### PR TITLE
coll/han: Implement MPI_Alltoallv in han using SMSC and XPMEM

### PR DESCRIPTION
**Pull request by lrbison**
Extension of the previous han MPI_Alltoall algorithm, this change adds MPI_Alltoallv to han for a hierarchy-aware algorithm which uses XPMEM via the SMSC module in order to directly read data from ranks on the same host.

The provides significant speed-up over the basic implementation when small messages are used, as many messages can be coalesced and packed into fewer sends.

Introduces MCA parameters:
 - coll_han_alltoallv_smsc_avg_send_limit
 - coll_han_alltoallv_smsc_noncontig_limit

### CODE REVIEW

**Comment by lrbison** [comment]
Here is some motivating data on a 32-host EFA cluster of hpc7g.16xlarge instances:


image


In particular this change makes small message sizes _much_ faster for Alltoallv.  The most improvement should be expected when messages are small, and there are many ranks on a single host.

Consider the scenario of 2 hosts (A and B) with 64 ranks per host transferring 4 byte messages.  In total the collective will send 128\*128 = 16K messages.  The existing basic algorithm will cause 64\*64=4K local messages to be sent via SM on both host A and B, and another 4K from host A and the last 4K from host B.  After performing an Allgather, this algorithm sends 1 message per rank per host, so host A sends 64 messages to host B and vice versa (and the other 64 messages are "send-to-self" sends).  This results in significantly less overhead and contention when compared to the basic algorithm.

However once the message size is large enough, the extra copy incurred by the pack-and-send method used starts to cause  mixed results, so I have implemented a size-dependent fallback to the previous implementation.  In fact, I find that the size I picked is close to the right answer for small clusters, but could still leave some performance on the table for large clusters (like the 32-node one shown above, which could gain 30% performance between 8KB and 64KB message sizes).

File: ompi/mca/coll/han/coll_han_alltoall.c [hunk 1]
```diff
@@ -31,7 +31,11 @@
 #include "opal/mca/rcache/rcache.h"
 #include "ompi/mca/osc/base/base.h"
 
-
+#if 1
+#define DEBUG_PRINT(...) do{ fprintf( stdout, __VA_ARGS__ ); } while( false )
+#else
+#define DEBUG_PRINT(...) do{ } while ( false )
+#endif
```
Comment (lrbison) [inline ompi/mca/coll/han/coll_han_alltoall.c:38]: Oops, I will remove these in future revision.

File: ompi/mca/coll/han/coll_han_alltoallv.c [hunk 2]
```diff
@@ -0,0 +1,878 @@
...
+    accurate.  We could be precisely correct and compute need_buffers for every
+    rank's count, but that could be a lot of iteration.  Just use 2 and assume
+    some rank will be sending more than 1 item, and adjust packed size
+    accordingly. */
+    const int count_for_convertor = 2;
+
+    OBJ_CONSTRUCT( &convertor, opal_convertor_t );
+    rc = opal_convertor_copy_and_prepare_for_recv(ompi_mpi_local_convertor,
+        &rdtype->super, count_for_convertor, rbuf, 0, &convertor);
+    bufs_on_device |= opal_convertor_on_device(&convertor);
```
Comment (bosilca) [inline ompi/mca/coll/han/coll_han_alltoallv.c:522]: why `|`?

File: ompi/mca/coll/han/coll_han_alltoallv.c [hunk 3]
```diff
@@ -0,0 +1,878 @@
...
+        memory cannot be registered with ibv_reg_mr.
+     3. Contiguous buffers.  The exception to #2 above is if we can't post our
+        sends/recvs in one large block to the NIC.  For these non-contiguous
+        datatypes, our packing algorithm is better because (a) we re-use our
+        buffers from a free-list which can remain registered to the NIC, and (b)
+        we pipeline multiple sends at once (c) we can pack more data into each
+        send than the network layer because we have mmap'ed all local ranks and
+        understand their datatypes.
+
+    The allgather adds about 50 usec on 256 ranks across 4 nodes, but for large
```
Comment (bosilca) [inline ompi/mca/coll/han/coll_han_alltoallv.c:507]: This comment talk about an allgather but the code uses an allreduce. What did I miss ?

File: ompi/mca/coll/han/coll_han_alltoallv.c [hunk 4]
```diff
@@ -0,0 +1,878 @@
...
+
+    {
+        static int attach_done=0;
+        if (attach_done) goto resume;
+        char *dbg_rank = getenv("DEBUG_RANK");
+        if (!dbg_rank) goto resume;
+        if (atoi(dbg_rank) != ompi_comm_rank(MPI_COMM_WORLD)) goto resume;
+        volatile int i = 0;
+        char hostname[256];
+        sleep(1);
```
Comment (bosilca) [inline ompi/mca/coll/han/coll_han_alltoallv.c:599]: Sleep(1) ? gethostname ? What's the scope of all these ? Why is this necessary on each call ?

**Review by bosilca — COMMENTED** [review]
I have left quite a few comments in the PR. In addition to them I would like to understand why it is necessary to heavily manipulate datatypes ? If I understand the code you are aggregating on the source for the same destination, so overall from the same architecture to the same architecture. Thus, you can cut most of the datatype use and simply rely on packed data. Or I might have missed something in the algorithm, in which case please help me understand.

File: ompi/mca/coll/han/coll_han_alltoallv.c (hunk 3 above)
Comment (lrbison) [reply ompi/mca/coll/han/coll_han_alltoallv.c]: My mistake: should say allreduce.  Addressed.

File: ompi/mca/coll/han/coll_han_alltoallv.c (hunk 2 above)
Comment (lrbison) [reply ompi/mca/coll/han/coll_han_alltoallv.c]: I've changed it to just assignment to be more clear.

File: ompi/mca/coll/han/coll_han_alltoallv.c (hunk 4 above)
Comment (lrbison) [reply ompi/mca/coll/han/coll_han_alltoallv.c]: Oops.  That's debug code that slipped past me.  I've removed it.

File: ompi/mca/coll/han/coll_han_alltoall.c [hunk 5]
```diff
@@ -331,6 +335,22 @@ int mca_coll_han_alltoall_using_smsc(
                 /* pack the data directly into local leader's sendbuf */
                 packed_size_tmp = packed_size;
                 rc = opal_convertor_pack(&convertor, &iov, &iov_count, &packed_size_tmp);
+
```
[the remark this answers is not in the document]
Comment (lrbison) [reply ompi/mca/coll/han/coll_han_alltoall.c]: There was no intent to update alltoall.c here.  I've removed it.

**Comment by lrbison** [comment]
> I would like to understand why it is necessary to heavily manipulate datatypes

@bosilca 

Sure.  The motivation is to have every rank on one host be able to mmap local sendbufs and therefore bypass a bunch of local messaging, and coalesce sends into big enough chunks to max out the NIC.  I am not attempting to change data types, just to share enough of the datatype description among the ranks in the low communicator so that `opal_convertor_pack` can be called.

I chose to assign the packing work to the sending rank, which means that the sending rank must understand how to pack not just its own send datatype, but also the other low-comm ranks' send-datatypes.  To do this I leaned on some code you and Joseph shared a few months back.  Using that I can serialize the opal datatypes of each rank and share them among the low-comm, and deserialize each of them on the other side without any modification.  Now that the sending rank understand the data type of all low-comm ranks, it packs the data from each of them into a series of buffers, and sends those buffers as MPI_PACKED datatype.

On the receiving side, the data is unpacked using only the user-provided recv datatype since all data received is of the same type.

Some alternatives:
 - Rather than have the sender do packing work, instead expose the send buffers and make it the job of the originating ranks to pack their data into the sender's waiting bounce buffers.  This is similar to "push mode" in my alltoall PR.  It requires additional synchronization which will be even more costly here where I use many small in-flight buffers rather than one big buffer as in alltoall.
 - `send()` the data between low-comm ranks and ignore SMSC.  But this incurs lots of additional message overhead which I found to be costly, and the intermediate rank must still know how many bytes it must receive especially if we intend to post any of those recvs in parallel.
 - `send()` the data directly to where it goes.  That's just `coll_basic_alltoallv`.

**Comment by bosilca** [comment]
> Rather than have the sender do packing work, instead expose the send buffers and make it the job of the originating ranks to pack their data into the sender's waiting bounce buffers. This is similar to "push mode" in my alltoall PR. It requires additional synchronization which will be even more costly here where I use many small in-flight buffers rather than one big buffer as in alltoall.

Yes, but instead of sequentialization the datatype creation and packing you can get them working in parallel. Technically, each process gets the packed size of the local datatype (mixed with the count), then do a scan on the local comm to get the starting point, and then all local processes in parallel do the pack onto the target architecture. I understand that all processes will have to do this for a different set of data, but at least this will save the time for creating a bunch of datatypes on each process.

Here I assume that even if the job is heterogeneous, all processes on the same node are homogeneous. This makes the convertor identical on all processes that participate to a local packing and then unpacking. A similar approach can be used on the receiver side, allowing all processes to unpack in parallel.

**Comment by lrbison** [comment]
I probably should write up a diagram to help visualize how the data movement happens.  For now here is some text:  Let's consider M ranks per host.  The algorithm is executed in rounds, with as many rounds as hosts.  Let us consider a particular round where the host with rank 0 is sending to another host which has rank X, where X=M*`<integer>`.  Simultaneously Rank X will be packing and sending back to rank 0.

> Yes, but instead of sequentialization the datatype creation and packing...

Data movement is already in parallel.  While rank 0 is packing data from 0:M to send to rank X+0, rank K is also packing data from rank 0:M to send to rank X+K (for K on the same host as 0, and X being the first rank on some remote host).

You are right though that we are making a bunch of extra opal datatypes.  However I had some debug code that skipped all the actual packing and sending, and the rest of the overhead was quite small (I don't have the numbers), so I concluded that overhead is not worth concerning myself over.

On the receiving in parallel: The sender (rank K) has already done the gather work across ranks, collecting data which ranks 0-M need to get to the receiver (X+K).  The receiver therefore only needs to unpack and "scatter" it across it's own recv buffer, according to the offsets the user provided.

>  Technically, each process gets the packed size of the local datatype (mixed with the count), then do a scan on the local comm to get the starting point

This is true, but made more tricky by the fact that my implementation doesn't need to allocate enough packing buffers to hold _all_ the data, instead it continually reuses a set of 8 buffers of 128KB each.  This means we'll have to notify local ranks each time a new buffer is ready for packing.

> Here I assume that even if the job is heterogeneous, all processes on the same node are homogeneous

Agreed.  But does that mean that `desc_opt` is necessarily available?

**Comment by lrbison** [comment]
Here is a quick picture:

image

**Comment by bosilca** [comment]
What happened to `Rank 2` ? The unlucky number if supposed to be `13` !?!?!

**Comment by bosilca** [comment]
> Agreed. But does that mean that desc_opt is necessarily available?

Yes, if an optimized description exists then `opt_desc` will point to it , otherwise 'opt_desc` will point back to `desc`. We can argue about what an `optimized description` is, for me it anything that will lead to more regular or lesser memory copies.

**Review by bosilca — APPROVED** [review]
